### PR TITLE
Action: Staging Deploy

### PR DESCRIPTION
**Pull request by zwolf**
- Sets up a staging deploy via Github Action for FEM. A few idiosyncrasies force this to be a bit custom.
- Image builds will continue in both Hub and GHCR until the Jenkinsfile is removed.
- The contentful secrets need to be added to the repo (or org, if more widely used) by someone who knows what they are (I didn't find them in Passbolt).
- Not currently using shared workflows because of
    - build-args. Is this a requirement? Should the ci-cd version take build-args as a `with:` param?
    - Separation of `-staging` and `-production` repos in the container registry. Is that necessary? It’s deploying a tagged hash either way, this just necessitates rebuilding & pushing the image to the production repo at deploy. Also means this can’t use the shared ci-cd deploy_app workflow, but that might be true regardless since there’s two deployments in there and the “deploy necessary?” check wouldn’t account for that.

 

## How to Review
_Helpful explanations that will make your reviewer happy:_
- What Zooniverse project should my reviewer use to review UX?
- What user actions should my reviewer step through to review this PR?
- Which storybook stories should be reviewed?
- Are there plans for follow up PR’s to further fix this bug or develop this feature?

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## Bug Fix
- [ ] The PR creator has listed user actions to use when testing if bug is fixed
- [ ] The bug is fixed
- [ ] Unit tests are added or updated

## New Feature
- [ ] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature
- [ ] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

## Refactoring
- [ ] The PR creator has described the reason for refactoring
- [ ] The refactored component(s) continue to work as expected

## Maintenance
- [ ] If not from dependabot, the PR creator has described the update (major, minor, or patch version, changelog)

### CODE REVIEW

**Comment by eatyourgreens** [comment]
There’s a GitHub Action that builds the NextJS apps during CI, which means the Contentful secrets must already be in this repo. 

We do need separate images for staging and production. I think because each has its own Sentry setup at build time, but I can’t remember off the top of my head. I do remember that we had a single image until we added Sentry logging.

**Comment by eatyourgreens** [comment]
Here are the NextJS build secrets, for production at least. I _think_ staging uses the same Contentful tokens, just different asset prefixes. 

Actually, I should double check these asset prefixes, now that we serve assets via the CDN. This test workflow could be out-of-date. 


https://github.com/zooniverse/front-end-monorepo/blob/9530ecd8dd125a517f07101dd736689d3529c4c8/.github/workflows/ci-tests.yml#L58-L65

**Comment by zwolf** [comment]
The contentful keys are repo secrets with the names I assumed in this PR they'd have, but they should also be in Passbolt. The asset prefixes were copied along with the rest of the env vars passed as build-args to docker in the Jenkinsfile.

**Comment by eatyourgreens** [comment]
I remembered why we have a staging image and a production image: the static JS bundles that webpack builds during `next build` are different in staging and production.

File: .github/workflows/deploy_staging.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        push: true
+        tags: |
+          ghcr.io/zooniverse/front-end-monorepo-staging:latest
+          ghcr.io/zooniverse/front-end-monorepo-staging:${{ github.sha }}
+        build-args: |
+          APP_ENV="staging"
+          COMMIT_ID=${{ github.sha }}
+          CONTENTFUL_ACCESS_TOKEN=${{ secrets.CONTENTFUL_ACCESS_TOKEN }}
+          CONTENTFUL_SPACE_ID=${{ secrets.CONTENTFUL_SPACE_ID }}
+          CONTENT_ASSET_PREFIX=https://fe-content-pages.preview.zooniverse.org
```
Comment (eatyourgreens) [inline .github/workflows/deploy_staging.yml:35]:
```suggestion
          CONTENT_ASSET_PREFIX=https://frontend.preview.zooniverse.org/about
```

File: .github/workflows/deploy_staging.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+          SENTRY_CONTENT_DSN=https://1f0126a750244108be76957b989081e8@sentry.io/1492498
+          PROJECT_ASSET_PREFIX=https://fe-project.preview.zooniverse.org
```
Comment (eatyourgreens) [inline .github/workflows/deploy_staging.yml:37]:
```suggestion
          PROJECT_ASSET_PREFIX=https://frontend.preview.zooniverse.org/projects
```

**Review by eatyourgreens — COMMENTED** [review]
These are the correct values for the asset prefixes. I completely missed these in #2541.

**Comment by eatyourgreens** [comment]
The Contentful API keys are in our Contentful account, under Settings > API Keys.

Screenshot of the Contentful Settings menu, including the API Keys option.

**Review by eatyourgreens — APPROVED** [review]
Looks good to me. I'm not sure if those asset prefixes are used to build the static JS bundles, but it can't hurt to use values that serve static assets via the CDN.

**Pull request merged**